### ISSUE TITLE
ESIPP [Slow] e2e tests failing on large cluster release-blocking job

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#52495]
The following that have been consistently failing on our 5k-node tests (https://k8s-testgrid.appspot.com/google-gce-scale#gce-scale-correctness):
- ESIPP [Slow] should handle updates to ExternalTrafficPolicy field
- ESIPP [Slow] should only target nodes with endpoints
- ESIPP [Slow] should work for type=LoadBalancer
- ESIPP [Slow] should work from pods
- Services should be able to create an internal type load balancer [Slow]
- Services should only allow access from service loadbalancer source ranges [Slow]

This is part of the correctness suite and large cluster tests are release-blocking.

cc @kubernetes/sig-network-bugs @kubernetes/sig-scalability-misc

**Comment by shyamjvs** [comment]
@bowei Assigning this to you. Feel free to redirect as apt.

**Comment by shyamjvs** [comment]
Ref https://github.com/kubernetes/kubernetes/issues/51718

cc @nicksardo

**Comment by shyamjvs** [comment]
Btw I'm seeing that almost all of those tests failed with the same kind of reason - `Timeout waiting for service "<xyz>" to have a load balancer`
@nicksardo Does it look like the same as https://github.com/kubernetes/kubernetes/issues/51714 to you?

**Comment by nicksardo** [comment]
Yes, I think we can close #51714 as it doesn't seem to be an internal LB specific issue.

**Comment by shyamjvs** [comment]
We can probably close that issue. So is it the problem with load balancer not getting created with custom subnets?
Btw - is there a fix on flight for this already (I vaguely remember seeing a pr around it)?

**Comment by nicksardo** [comment]
The custom subnet problem was fixed. The latest test run shows the kube controller is being sent the subnet name.  I don't know why the LBs are timing out - still trying to make sense of the kube-controller-manager log.

**Comment by nicksardo** [comment]
Looking at the project's GCP activity log, it takes ~13 minutes for the 'add instance to target pool' operation to complete. I'm wondering if the test timed out before GCP could catch up.  It also seems that the service controller didn't even attempt to create the internal LB.

**Comment by nicksardo** [comment]
@shyamjvs is this the first time these LB tests have been run in a scalability suite?

**Comment by nicksardo** [comment]
It also appears that the forwarding rule creation operations are taking ~18 minutes at this scale. Not surprising that all the load balancers are timing out and the service controller never got around to creating the ILB.

**Comment by shyamjvs** [comment]
To add some clarity:
- We've been running correctness tests on 5000-node clusters for about 2 releases now and they're defined by '--ginkgo.skip=\[Serial\]|\[Disruptive\]|\[Flaky\]|\[Feature:.+\]'. So pretty much all suites except those ones are covered by it. And note that this also includes [Slow] suite. It seems like these ESIPP tests are new as I don't remember seeing them a while ago.
- Also those tests are on GCE and not GKE (because we don't want to block OSS release based on GKE)
- We're also running correctness tests on a slightly lower scale (i.e 2000 nodes) and there the ESIPP suite seems to have passed 2/3 times from the recent runs - https://k8s-testgrid.appspot.com/google-gce-scale#gce-large-correctness . So it could likely be that the test might need to increase timeout based on cluster-size (just a guess).

Also, did you find anything interesting in the controller-manager logs (http://gcsweb.k8s.io/gcs/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/12/artifacts/gce-scale-cluster-master/)?
I tried taking a quick look and extracted the following logs that seem to be most interesting:

```
I0914 07:47:26.089108       7 event.go:218] Event(v1.ObjectReference{Kind:"ReplicationController", Namespace:"e2e-tests-services-l4bxj", Name:"lb-sourcerange", UID:"f3719d6d-9920-11e7-bcbc-42010a8e0002", APIVersion:"v1", ResourceVersion:"1372301", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: lb-sourcerange-n5wh9
I0914 07:47:35.461014       7 event.go:218] Event(v1.ObjectReference{Kind:"ReplicationController", Namespace:"e2e-tests-services-fp9fg", Name:"lb-internal", UID:"f90e5c84-9920-11e7-bcbc-42010a8e0002", APIVersion:"v1", ResourceVersion:"1378095", FieldPath:""}): type: 'Normal' reason: 'SuccessfulCreate' Created pod: lb-internal-wpp7x
I0914 08:07:52.980323       7 replication_controller.go:597] Replication Controller has been deleted e2e-tests-services-fp9fg/lb-internal
I0914 09:35:24.212586       7 replication_controller.go:597] Replication Controller has been deleted e2e-tests-services-fp9fg/lb-internal
I0914 09:44:02.206758       7 replication_controller.go:597] Replication Controller has been deleted e2e-tests-services-fp9fg/lb-internal
I0914 09:47:45.983739       7 replication_controller.go:597] Replication Controller has been deleted e2e-tests-services-l4bxj/lb-sourcerange
I0914 09:47:51.093051       7 namespace_controller.go:171] Namespace has been deleted e2e-tests-services-l4bxj
I0914 09:48:01.024837       7 gce_loadbalancer_external.go:189] EnsureLoadBalancer(af72f6e8d992011e7bcbc42010a8e000(e2e-tests-services-l4bxj/lb-sourcerange)): creating firewall
I0914 09:48:19.846906       7 gce_loadbalancer_external.go:193] EnsureLoadBalancer(af72f6e8d992011e7bcbc42010a8e000(e2e-tests-services-l4bxj/lb-sourcerange)): created firewall
I0914 09:48:19.972425       7 gce_loadbalancer_external.go:202] Target pool af72f6e8d992011e7bcbc42010a8e000 for Service e2e-tests-services-l4bxj/lb-sourcerange doesn't exist
I0914 09:48:20.361131       7 gce_loadbalancer_external.go:527] Creating targetpool af72f6e8d992011e7bcbc42010a8e000 with 1 healthchecks
I0914 09:48:54.388339       7 gce_loadbalancer_external.go:280] EnsureLoadBalancer(af72f6e8d992011e7bcbc42010a8e000(e2e-tests-services-l4bxj/lb-sourcerange)): created health checks k8s-573aeda2986fb435-node for target pool
I0914 09:48:54.388391       7 gce_loadbalancer_external.go:285] EnsureLoadBalancer(af72f6e8d992011e7bcbc42010a8e000(e2e-tests-services-l4bxj/lb-sourcerange)): created initial target pool (now updating with 4800 hosts)
I0914 10:01:58.361770       7 gce_loadbalancer_external.go:294] EnsureLoadBalancer(af72f6e8d992011e7bcbc42010a8e000(e2e-tests-services-l4bxj/lb-sourcerange)): updated target pool (with 4800 hosts)
I0914 10:01:58.361816       7 gce_loadbalancer_external.go:298] EnsureLoadBalancer(af72f6e8d992011e7bcbc42010a8e000(e2e-tests-services-l4bxj/lb-sourcerange)): creating forwarding rule, IP 35.196.224.253 (tier: Premium)
I0914 10:20:44.445532       7 gce_loadbalancer_external.go:307] EnsureLoadBalancer(af72f6e8d992011e7bcbc42010a8e000(e2e-tests-services-l4bxj/lb-sourcerange)): created forwarding rule, IP 35.196.224.253
E0914 10:20:48.130150       7 service_controller.go:749] Failed to process service. Retrying in 5s: Failed to persist updated status to apiserver, even after retries. Giving up: Not persisting update to service 'e2e-tests-services-l4bxj/lb-sourcerange' that has been changed since we received it: Operation cannot be fulfilled on services "lb-sourcerange": StorageError: invalid object, Code: 4, Key: /registry/services/specs/e2e-tests-services-l4bxj/lb-sourcerange, ResourceVersion: 0, AdditionalErrorMsg: Precondition failed: UID in precondition: f72f6e8d-9920-11e7-bcbc-42010a8e0002, UID in object meta: 
I0914 10:20:48.130194       7 service_controller.go:736] Service has been deleted e2e-tests-services-fp9fg/lb-internal
E0914 10:20:48.130206       7 service_controller.go:757] Failed to process service. Not retrying: Service e2e-tests-services-fp9fg/lb-internal not in cache even though the watcher thought it was. Ignoring the deletion.
I0914 10:20:48.130218       7 service_controller.go:736] Service has been deleted e2e-tests-dns-dqhqt/dns-test-service
E0914 10:20:48.130225       7 service_controller.go:757] Failed to process service. Not retrying: Service e2e-tests-dns-dqhqt/dns-test-service not in cache even though the watcher thought it was. Ignoring the deletion.
I0914 10:20:48.131333       7 service_controller.go:736] Service has been deleted e2e-tests-dns-dqhqt/test-service-2
```

There are 2 things that seem wrong to me here:
- Why is it updating with 4800 instances when there are 5000?
- By the time the service_controller got to processing the service and it was on the way, the test timed out and the service deleted.

**Comment by shyamjvs** [comment]
Unfortunately I'm a bit occupied by few other things at the moment and couldn't look deeper. Can this test be fixed somehow in the short-term - by increasing timeout based on cluster size or something?

**Comment by MrHohn** [comment]
> It seems like these ESIPP tests are new as I don't remember seeing them a while ago.

Some of the ESIPP tests were moved into slow suite for 1.6 release. Ref: https://github.com/kubernetes/kubernetes/pull/44582

**Comment by MrHohn** [comment]
> Is there a total timeout on running correctness tests

It seems to be [1000 minutes for gce-scale-correctness](https://github.com/kubernetes/test-infra/blob/f4c3cf4798d22fd76fe4cf07050bf3f281140b57/jobs/config.json#L2847-L2864) and [600 minutes for gce-large-correctness](https://github.com/kubernetes/test-infra/blob/f4c3cf4798d22fd76fe4cf07050bf3f281140b57/jobs/config.json#L2384-L2401).

**Comment by MrHohn** [comment]
@shyamjvs Would be great if we keep the previous succeeded build logs for the correctness jobs.

From gubernator it seems we have logs only after 1.7 and they never passed:
- https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/?
- https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-large-correctness/?

**Comment by shyamjvs** [comment]
> It seems to be 1000 minutes for gce-scale-correctness and 600 minutes for gce-large-correctness.

@MrHohn @nicksardo That's right. And currently the gce-scale-correctness job finishes in about 5-7hrs. So I don't think it would be a problem to bump the timeout for your e2e by a few minutes (if you feel that would fix it).

> Would be great if we keep the previous succeeded build logs for the correctness jobs.

I agree. The problem is we only recently (about a month or so ago) automated our scalability tests to run as part of CI. Previously we were running them manually, just prior to the release. On the good side, we'd be having them from now on :)

**Comment by MrHohn** [comment]
> @mrhohn @nicksardo That's right. And currently the gce-scale-correctness job finishes in about 5-7hrs. So I don't think it would be a problem to bump the timeout for your e2e by a few minutes (if you feel that would fix it).

Ack, will give it a shot.

/assign

**Comment by MrHohn** [comment]
Humm, it seems we already set a large timeout (2 hours) for LB tests when running on big cluster.

https://github.com/kubernetes/kubernetes/blob/bb50086b8ffb4289e085fce637a5cd01f8ef4296/test/e2e/framework/service_util.go#L73

https://github.com/kubernetes/kubernetes/blob/bb50086b8ffb4289e085fce637a5cd01f8ef4296/test/e2e/network/service.go#L1433-L1435

But [we still hit timeout](https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/12):
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/network/service.go:790
Sep 14 11:13:04.471: Could not reach HTTP service through 35.196.237.29:81 after 2h0m0s: timed out waiting for the condition
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/service_util.go:842
```

**Comment by MrHohn** [comment]
Regarding my previous comment, I just noticed that we run tests in parallel for the correctness jobs. Given this, we would probably try to create 7 LBs at the same time, at the worse case it might take (17 mins ( forwarding rule creation) + 14 mins (target pool creation/addition) + 1~2 mins (firewall rules, healthcheck etc.)) * 7 = larger than 3.5 hours to finish, and test will certainly timeout.

Although, some of the LB tests should have passed given that their timeout is 2 hours. I'm still tracing down what exactly happened. The logs posted in https://github.com/kubernetes/kubernetes/issues/52495#issuecomment-329746924 do look a bit suspicious, it maybe some issues related to how updates is handled in service controller / gce controller.

On the other hand, consider most of the CPU time in service controller is wasted on waiting for GCE operation to finish, I think it is reasonable to bump `--concurrent-service-syncs` for kube-controller-manager for large cluster, which represents the number of services that are allowed to sync concurrently. By default it is 1 and is clearly not enough in this case.

**Comment by shyamjvs** [comment]
> Regarding my previous comment, I just noticed that we run tests in parallel for the correctness jobs. Given this, we would probably try to create ....

Makes sense - though like you later said, at least some of those should have passed (even if they all ran serially). Btw - you can try manually creating a large cluster and see what's happening live (we have quota in one of our projects (k8s-scale-testing) specially meant for such manual scalability experiments).

> I think it is reasonable to bump --concurrent-service-syncs for kube-controller-manager for large cluster, which represents the number of services that are allowed to sync concurrently.

Sg - though I'm not very sure if we want to do it as it might diverge our test setup from production setup. Maybe we can do it just for this test's sake?
@gmarek @wojtek-t - Any comments?

**Comment by shyamjvs** [comment]
Removing this from release-blocker as it seems to be mostly related to operations on cloud-provider side. Though it is still important to be fixed.

**Comment by MrHohn** [comment]
Also some explanations for the error messages posted on https://github.com/kubernetes/kubernetes/issues/52495#issuecomment-329746924:
```
I0914 10:20:48.130194       7 service_controller.go:736] Service has been deleted e2e-tests-services-fp9fg/lb-internal
E0914 10:20:48.130206       7 service_controller.go:757] Failed to process service. Not retrying: Service e2e-tests-services-fp9fg/lb-internal not in cache even though the watcher thought it was. Ignoring the deletion.
I0914 10:20:48.130218       7 service_controller.go:736] Service has been deleted e2e-tests-dns-dqhqt/dns-test-service
E0914 10:20:48.130225       7 service_controller.go:757] Failed to process service. Not retrying: Service e2e-tests-dns-dqhqt/dns-test-service not in cache even though the watcher thought it was. Ignoring the deletion.
```

This basically happens when a service is created and later be deleted, and when the service controller isn't responsive enough --- it doesn't response to a service creation until after the service is deleted. Because we put a workqueue in between service watcher and service processing, a "creation+deletion" combination will be transferred into a single deletion in this case. Graphically:

![service_processing](https://user-images.githubusercontent.com/8681801/30614914-eb20fb9e-9d41-11e7-91f2-782ffeff0a2e.jpg)

In short, this error message is as expected.

Also, an interesting opposite case for this is a service being deleted and created (with the same name) in a short period, and that caused issue before: https://github.com/kubernetes/kubernetes/issues/43730.

**Comment by MrHohn** [comment]
> Although, some of the LB tests should have passed given that their timeout is 2 hours.

More explanation for this. It seems we are using the working queue in a FIFO fashion for service update:

https://github.com/kubernetes/kubernetes/blob/dc460c3e15651d7e3527337c7997ab7f956ad6b1/staging/src/k8s.io/client-go/util/workqueue/queue.go#L86-L106

https://github.com/kubernetes/kubernetes/blob/dc460c3e15651d7e3527337c7997ab7f956ad6b1/staging/src/k8s.io/client-go/util/workqueue/queue.go#L117-L139
 

As a result of how the e2e tests proceed, service deletion will mostly be queued after many of the LB service creations. This explains why some of the tests failed on waiting for LB to be created while others failed on waiting for LB to be disappear.

Ref: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/13

**Comment by shyamjvs** [comment]
Seems like the tests are still timing out - https://k8s-testgrid.appspot.com/google-gce-scale#gce-scale-correctness

@MrHohn Sorry I've had some context loss. I remember that we were being bottlenecked on service queue processing due to cloud operations - so you increased the parallelism.
- Are we just waiting for cloud-provider side issues to be resolved?
- Will increasing the timeout for e2es work (I think we were worried about worst case time being high if all tests run serially, but that should be unlikely given ginkgo parallelism of 30 and other tests finishing soon)?

**Comment by shyamjvs** [comment]
cc @wojtek-t

**Comment by MrHohn** [comment]
> Are we just waiting for cloud-provider side issues to be resolved?

Nope, I believe the limited throughput on the service controller side also needs to be resolved.

> Will increasing the timeout for e2es work (I think we were worried about worst case time being high if all tests run serially, but that should be unlikely given ginkgo parallelism of 30 and other tests finishing soon)?

Test may pass if we increase the timeout for each LB e2e test (to 4 or 5 hours? We already set 2 hours...) regardless, though it seems unreasonable if service controller takes 4~5 hours to provision cloud resources for a LB service. 

Dug into the logs a bit and found service controller [still process service update serially](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/18/artifacts/gce-scale-cluster-master/kube-controller-manager.log) after https://github.com/kubernetes/test-infra/pull/4899. Turned out I forgot to pipe `CONCURRENT_SERVICE_SYNCS` into `kube-env` (https://github.com/kubernetes/kubernetes/pull/52684) so what we had set have no effects :(

I sent #54033 to properly pass `CONCURRENT_SERVICE_SYNCS` onto master.

BTW I saw below error from one of the scale test (I forgot which). Does ILB only support up to 2000 instances? May be we should simply disable ILB test in scale CI? @nicksardo 
```
E1010 08:02:17.299265       7 service_controller.go:754] Failed to process service. Retrying in 2m40s: Failed to ensure load balancer for service e2e-tests-services-tmk49/lb-internal: googleapi: Error 403: Exceeded limit 'MAX_INSTANCES_IN_INSTANCE_GROUP' on resource 'k8s-ig--5701e438db2b4e8e'. Limit: 2000.0, limitExceeded
I1010 08:02:17.299386       7 event.go:218] Event(v1.ObjectReference{Kind:"Service", Namespace:"e2e-tests-services-tmk49", Name:"lb-internal", UID:"e7f04fc1-ad90-11e7-990f-42010a8e0002", APIVersion:"v1", ResourceVersion:"1102746", FieldPath:""}): type: 'Warning' reason: 'CreatingLoadBalancerFailed' Error creating load balancer (will retry): Failed to ensure load balancer for service e2e-tests-services-tmk49/lb-internal: googleapi: Error 403: Exceeded limit 'MAX_INSTANCES_IN_INSTANCE_GROUP' on resource 'k8s-ig--5701e438db2b4e8e'. Limit: 2000.0, limitExceeded
``

**Comment by shyamjvs** [comment]
Sounds good. Btw - we can just use `CONTROLLER_MANAGER_TEST_ARGS` to pass the parallelism flag instead of creating/piping a new env var just for that flag imo.

**Comment by nicksardo** [comment]
Thanks for pointing that out @MrHohn, but that doesn't look ILB specific. "Exceeded limit 'MAX_INSTANCES_IN_INSTANCE_GROUP' on resource 'k8s-ig--5701e438db2b4e8e'. Limit: 2000.0, limitExceeded"  This affects GCE Ingress as well, which I believe we're already not testing. Yes, we should probably disable the ILB test in scale CI.

**Comment by wojtek-t** [comment]
> regardless, though it seems unreasonable if service controller takes 4~5 hours to provision cloud resources for a LB service.

It's not 4-5 hours. A bunch of tests are fitting 2 hours, so my feeling is that in those failures we are not missing that much.
But the most important point is that the test itself is not just setting up LB - it's doing some manipulations on it too. So the single operation is visibly shorter.

That said - I completely agree that it's still unreasonably long - we should work with GCE networking team to have it improved.

**Comment by shyamjvs** [comment]
@MrHohn Seems like some of those e2es are in a better state after change in service-controller parallelism. However, there are still a few failing (https://k8s-testgrid.appspot.com/sig-release-master-blocking#gce-scale-correctness):

- ESIPP [Slow] should handle updates to ExternalTrafficPolicy field
- ESIPP [Slow] should work from pods
- Services should be able to change the type and ports of a service [Slow]

Do we have a way to fix them?

**Comment by shyamjvs** [comment]
cc @porridge

**Comment by MrHohn** [comment]
Took a look those failing tests, from service controller perspective I'd say it is acting normally, but in the middle of the test it hit a `googleapi: Error 503` following with a lot of `resourceNotReady` failures last for 1.5 hours.

From https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-scale-correctness/22/artifacts/gce-scale-cluster-master/kube-controller-manager.log:
```
# Began with a 503 error.
I1026 08:54:02.890483       7 gce_loadbalancer_external.go:373] ensureExternalLoadBalancerDeleted(aaadc336eba2111e79b7042010a8e000(e2e-tests-esipp-9z9pl/external-local)): Deleting IP address.
E1026 09:24:12.920412       7 gce_op.go:88] GCE operation failed: googleapi: Error 503: Code: '1582556575436094169'
I1026 09:24:16.608356       7 gce_loadbalancer_external.go:129] ensureExternalLoadBalancer(ad239592aba2211e79b7042010a8e000(e2e-tests-services-58xks/lb-sourcerange)): Released static IP 35.190.129.220.
E1026 09:24:16.608447       7 service_controller.go:773] Failed to process service e2e-tests-services-58xks/lb-sourcerange. Retrying in 5s: Failed to ensure load balancer for service e2e-tests-services-58xk     s/lb-sourcerange: failed to create forwarding rule for load balancer (ad239592aba2211e79b7042010a8e000(e2e-tests-services-58xks/lb-sourcerange)): googleapi: Error 503: Code: '1582556575436094169'
...
# resourceNotReady started to occur.
I1026 09:47:15.128709       7 gce_loadbalancer_external.go:373] ensureExternalLoadBalancerDeleted(ad239592aba2211e79b7042010a8e000(e2e-tests-services-58xks/lb-sourcerange)): Deleting IP address.
E1026 09:47:30.593312       7 service_controller.go:773] Failed to process service e2e-tests-services-58xks/lb-sourcerange. Retrying in 5s: googleapi: Error 400: The resource 'projects/kubernetes-scale/regions/us-east1/forwardingRules/ad239592aba2211e79b7042010a8e000' is not ready, resourceNotReady
I1026 09:49:56.941054       7 gce_loadbalancer_external.go:153] ensureExternalLoadBalancer(a969c860dba2311e79b7042010a8e000(e2e-tests-esipp-sd5cq/external-local)): Ensured IP address 35.196.32.59 (tier: Premium).
I1026 09:50:01.132005       7 gce_loadbalancer_external.go:129] ensureExternalLoadBalancer(a969c860dba2311e79b7042010a8e000(e2e-tests-esipp-sd5cq/external-local)): Released static IP 35.196.32.59.
I1026 09:50:01.299140       7 service_controller.go:286] Deleting existing load balancer for service e2e-tests-esipp-tnldv/external-local that no longer needs a load balancer.
I1026 09:50:01.299245       7 gce_loadbalancer_external.go:377] ensureExternalLoadBalancerDeleted(a922a1c1bba2111e79b7042010a8e000(e2e-tests-esipp-tnldv/external-local)): Deleting forwarding rule.
I1026 09:50:01.299314       7 gce_loadbalancer_external.go:359] ensureExternalLoadBalancerDeleted(a922a1c1bba2111e79b7042010a8e000(e2e-tests-esipp-tnldv/external-local)): Deleting firewall rule.
I1026 09:50:01.299367       7 gce_loadbalancer_external.go:373] ensureExternalLoadBalancerDeleted(a922a1c1bba2111e79b7042010a8e000(e2e-tests-esipp-tnldv/external-local)): Deleting IP address.
E1026 09:50:01.525276       7 service_controller.go:773] Failed to process service e2e-tests-esipp-tnldv/external-local. Retrying in 10s: googleapi: Error 400: The resource 'projects/kubernetes-scale/regions/us-east1/forwardingRules/a922a1c1bba2111e79b7042010a8e000' is not ready, resourceNotReady
I1026 09:50:01.525356       7 service_controller.go:760] Service has been deleted e2e-tests-services-58xks/lb-sourcerange
I1026 09:50:01.525431       7 gce_loadbalancer_external.go:377] ensureExternalLoadBalancerDeleted(ad239592aba2211e79b7042010a8e000(e2e-tests-services-58xks/lb-sourcerange)): Deleting forwarding rule.
I1026 09:50:01.525494       7 gce_loadbalancer_external.go:359] ensureExternalLoadBalancerDeleted(ad239592aba2211e79b7042010a8e000(e2e-tests-services-58xks/lb-sourcerange)): Deleting firewall rule.
I1026 09:50:01.525698       7 gce_loadbalancer_external.go:373] ensureExternalLoadBalancerDeleted(ad239592aba2211e79b7042010a8e000(e2e-tests-services-58xks/lb-sourcerange)): Deleting IP address.
E1026 09:50:01.742593       7 service_controller.go:773] Failed to process service e2e-tests-services-58xks/lb-sourcerange. Retrying in 10s: googleapi: Error 400: The resource 'projects/kubernetes-scale/regions/us-east1/forwardingRules/ad239592aba2211e79b7042010a8e000' is not ready, resourceNotReady
E1026 09:50:17.003132       7 service_controller.go:773] Failed to process service e2e-tests-esipp-rcz42/external-local. Retrying in 5s: timed out waiting for the condition
I1026 09:50:17.209073       7 service_controller.go:286] Deleting existing load balancer for service e2e-tests-esipp-tnldv/external-local that no longer needs a load balancer.
I1026 09:50:17.209229       7 gce_loadbalancer_external.go:377] ensureExternalLoadBalancerDeleted(a922a1c1bba2111e79b7042010a8e000(e2e-tests-esipp-tnldv/external-local)): Deleting forwarding rule.
I1026 09:50:17.209260       7 gce_loadbalancer_external.go:359] ensureExternalLoadBalancerDeleted(a922a1c1bba2111e79b7042010a8e000(e2e-tests-esipp-tnldv/external-local)): Deleting firewall rule.
I1026 09:50:17.209295       7 gce_loadbalancer_external.go:373] ensureExternalLoadBalancerDeleted(a922a1c1bba2111e79b7042010a8e000(e2e-tests-esipp-tnldv/external-local)): Deleting IP address.
E1026 09:50:17.424496       7 service_controller.go:773] Failed to process service e2e-tests-esipp-tnldv/external-local. Retrying in 20s: googleapi: Error 400: The resource 'projects/kubernetes-scale/regions/us-east1/forwardingRules/a922a1c1bba2111e79b7042010a8e000' is not ready, resourceNotReady
...failures kept occurring for 1.5 hours and almost every LB related GCE operation failed. It seems like it eventually backed to normal but tests already timed out. 
```

It seems to me that GCE server failed to fulfill the operations demanded by service controller and acted abnormally during the test, potentially due to the huge instances number in target pools.

> Do we have a way to fix them?

Per findings above, there isn't much we can do on k8s side at this point.

**Comment by shyamjvs** [comment]
Bumping the priority as this is blocking 1.9-alpha release. Umbrella issue - #55194

**Comment by shyamjvs** [comment]
@MrHohn We're still seeing loadbalancer related failures and from your comment above, it seems like there's not much that can be done on the k8s side at this point.
In that case, I think the option we have is to disable `ESIPP [Slow]` in our 5k-node correctness job and add a note in the scalability-validation-report for the release ([e.g](https://github.com/kubernetes/features/blob/master/release-1.8/scalability_validation_report.md)) about this issue on GCE.

**Comment by shyamjvs** [comment]
As discussed offline with @MrHohn, we're going to disable the suite for huge clusters using an extra check.
Sent out #55275 for it with a TODO to enable later when cloud-provider side improvements are made.

**Comment by shyamjvs** [comment]
Ok.. With the ESIPP suite disabled, the job looks more greener - https://k8s-testgrid.appspot.com/sig-scalability-gce#gce-scale-correctness
There's still one networking e2e (not from ESIPP), that's failing:

```
Services should be able to change the type and ports of a service [Slow]
```

@MrHohn can we fix this or should we also skip it (given that it is also LB-related)?

**Comment by porridge** [comment]
Also, now that we are not tracking the result of these tests at large
scale, we will need to mention this in the scalability validation report
for 1.9, like we mentioned the regression in 1.8 at the bottom of
https://github.com/kubernetes/features/blob/master/release-1.8/scalability_validation_report.md
Can sig-testing or sig-gcp come up with an appropriate wording for this
remark?

**Comment by shyamjvs** [comment]
I discussed this briefly with @MrHohn offline.
IMO, sth like the following seems reasonable:

```
Correctness testing excluded 'ESIPP [Slow]' suite and a couple of other networking-related e2e tests dealing with (external?) loadbalancer, due to GCE-side issues.
Those tests are now skipped for clusters larger than 2k nodes, until further improvements are made. Ref: <this issue>
```

SG?

**Comment by MrHohn** [comment]
Re the failure on `Services should be able to change the type and ports of a service` (https://github.com/kubernetes/kubernetes/issues/52495#issuecomment-343185832): checked the service controller logs and found it hit the same issue as ESIPP, below `resourceNotReady` error is being constantly emitted:
```
E1109 11:48:30.729765       7 service_controller.go:779] Failed to process service e2e-tests-services-r7mqb/mutability-test. Retrying in 5m0s: googleapi: Error 400: The resource 'projects/kubernetes-scale/      regions/us-east1/forwardingRules/a16e06ca5c52a11e7ad2842010a8e000' is not ready, resourceNotReady
```

@shyamjvs Suggest to skip it for the same reason as for ESIPP tests.

**Comment by MrHohn** [comment]
cc @bowei for the scalability validation report being added.

**Comment by shyamjvs** [comment]
With the above PRs merged, we're not running those tests for the release-blocking 5k-node job anymore. Bumping down the priority of this issue.

**Comment by shyamjvs** [comment]
@MrHohn Assigning to you for the remaining part of this issue :)
Feel free to redirect it as apt.

**Comment by porridge** [comment]
Closing this in favor of #56138 so that this bug shows as closed on the umbrella issue.
/close

**Comment by shyamjvs** [comment]
@porridge There are a bunch of TODOs in the code that reference this issue. You can change those, but I think it's probably fine to let this issue be shown open under that umbrella.
There's a lot of prior context and information too under this bug.

**Comment by shyamjvs** [comment]
If needed you can change the issue title to indicate that it's not release-blocking anymore.

**Comment by porridge** [comment]
My intention is to be able to take a glance at the umbrella bug and judge
its progress by the number of green (unclosed) bug references.

**Comment by shyamjvs** [comment]
Yup.. I understand your concern. Just feel that bug duplication and context loss is a heavier price to pay than a green mark under the umbrella which would close soon after the release :)
